libyul/YulName.cpp

```diff
@@ -0,0 +1,112 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+// SPDX-License-Identifier: GPL-3.0
+
+#include <libyul/YulName.h>
+
+#include <libyul/Exceptions.h>
+
+#include <fmt/compile.h>
+
+namespace solidity::yul
+{
+
+YulNameRepository::YulNameRepository()
+{
+	auto const emptyName = defineName("");
+	yulAssert(emptyName == YulNameRepository::emptyName());
+}
+
+std::string_view YulNameRepository::requiredLabelOf(YulName const _name) const
+{
+	auto const label = labelOf(_name);
+	yulAssert(label.has_value(), "YulName currently has no defined label in the YulNameRepository.");
+	return label.value();
+}
+
+std::optional<YulNameRepository::YulName> YulNameRepository::nameOfLabel(std::string_view const _label) const
+{
+	auto const it = std::find(m_definedLabels.begin(), m_definedLabels.end(), _label);
+	if (it != m_definedLabels.end())
+	{
+		YulName labelName{static_cast<YulName>(std::distance(m_definedLabels.begin(), it))};
+		// mostly it'll be iota
+		if (!isDerivedName(labelName) && std::get<0>(m_names[static_cast<size_t>(labelName)]) == labelName)
+			return std::get<0>(m_names[static_cast<size_t>(labelName)]);
+		// if not iota, we have to search
+		auto itName = std::find(m_names.rbegin(), m_names.rend(), std::make_tuple(labelName, YulNameState::DEFINED));
+		if (itName != m_names.rend())
+			return std::get<0>(*itName);
+	}
+	return std::nullopt;
+}
+
+void YulNameRepository::generateLabels(std::set<YulName> const& _usedNames, std::set<std::string> const& _illegal)
+{
+	std::set<std::string> used;
+	std::set<YulName> toDerive;
+	for (auto const name: _usedNames)
+	{
+		if (!isDerivedName(name))
+		{
+			auto const label = labelOf(name);
+			yulAssert(label.has_value());
+			auto const [it, emplaced] = used.emplace(*label);
+			if (!emplaced || _illegal.count(*it) > 0)
+				// there's been a clash ,e.g., by calling generate labels twice;
+				// let's remove this name and derive it instead
+				toDerive.insert(name);
+		}
+	}
+
+	std::vector<std::tuple<std::string, YulName>> generated;
+	auto namesIt = _usedNames.begin();
+	for (size_t nameValue = 1; nameValue < m_names.size(); ++nameValue)
```

Comment:
I am a bit confused why you iterate over `m_names`. If I am reading this right, then what you really are doing are just iterating over `_usedNames`. But you need to know which name in `m_names` it represents, right?
But you know that, because `YulName` is basically exactly just an index into `m_names`, no?
What am I missing?

Reply:
This definitely should be simplified. It makes the loop harder to understand than it should be.

Also, it's only correct if `m_names` is ordered the same way a set of labels is. Can we actually assume that?

Reply:
BTW, in the current form it would be better to break the loop when you know that `namesIt == _usedNames.end()`.

Or you should assert that this is true at the end of the loop (which would be a way to safeguard against mismatched order).